lib/posix/pthread.c

```diff
@@ -135,12 +202,79 @@ int pthread_attr_setstack(pthread_attr_t *_attr, void *stackaddr, size_t stacksi
 	return 0;
 }
 
-static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+static bool pthread_attr_is_valid(const struct pthread_attr *attr)
+{
+	/*
+	 * FIXME: Pthread attribute must be non-null and it provides stack
+	 * pointer and stack size. So even though POSIX 1003.1 spec accepts
+	 * attrib as NULL but zephyr needs it initialized with valid stack.
+	 */
+	if (attr == NULL || attr->initialized == 0U || attr->stack == NULL ||
+	    attr->stacksize == 0) {
+		return false;
+	}
+
+	/* require a valid scheduler policy */
+	if (!valid_posix_policy(attr->schedpolicy)) {
+		return false;
+	}
+
+	/* require a valid detachstate */
+	if (!(attr->detachstate == PTHREAD_CREATE_JOINABLE ||
+	      attr->detachstate == PTHREAD_CREATE_DETACHED)) {
+		return false;
+	}
+
+	return true;
+}
+
+FUNC_NORETURN
+static void posix_thread_finalize(struct posix_thread *t, void *retval)
 {
-	void * (*fun_ptr)(void *) = arg3;
+	k_spinlock_key_t key;
+
+	/* move thread from run_q to done_q */
+	key = k_spin_lock(&q_lock);
+	sys_dlist_remove(&t->q_node);
+	sys_dlist_append(&done_q, &t->q_node);
+	t->qid = POSIX_THREAD_DONE_Q;
+	t->retval = retval;
+	k_spin_unlock(&q_lock, key);
 
-	fun_ptr(arg1);
-	pthread_exit(NULL);
+	k_thread_abort(k_current_get());
+	CODE_UNREACHABLE;
+}
+
+FUNC_NORETURN
+static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+{
+	int err;
+	void *arg;
+	k_spinlock_key_t key;
+	struct posix_thread *t;
+	pthread_barrier_t barrier;
+	void *(*fun_ptr)(void *arg);
+
+	arg = arg1;
```

Comment:
ah, now I understand how I missed the assignment to barrier -- these three assignments could be done as initializers instead. Might be a nice clean up. If you like things that way (which I do).

Reply:
OK